File: C#/Projects/SnomedQuery.Parser/RF2ConceptGroup.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SnomedQuery.Parser
{
  /// <summary>
  /// Group of concepts, all with same SNOMED id.
  /// </summary>
  public class RF2ConceptGroup : TRF2ItemGroup<RF2Concept>
  {
    /// <summary>
    /// List of all snomed descriptions for this concept.
    /// </summary>
    List<Int64> descriptionGroupIds = new List<Int64>();

    /// <summary>
    /// List of all snomed attribute concepts in which this concept is the
    /// source concept.
    /// </summary>
    List<Int64> sourceRelationshipIds = new List<Int64>();

    /// <summary>
    /// List of all snomed attribute concepts in which this concept is the
    /// destination concept.
    /// </summary>
    List<Int64> destinationRelationshipIds = new List<Int64>();

    /// <summary>
    /// Chain of is a parents.
    /// </summary>
    public List<Int64> IsAParents = new List<Int64>();

    /// <summary>
    /// Concept id for the class that this snomed 
    /// is an instance of.
    /// </summary>
    public Int64 ClassConceptId = -1;

    /// <summary>
    /// Top level node that this concept group is descended from.
    /// </summary>
    public Int64 TopLevelConceptId = -1;

    /// <summary>
    /// Lazy load Description groups.
    /// </summary>
    public RF2DescriptionGroup[] DescriptionGroups
    {
      get
      {
        // # Tested.
        if (this.descriptionGroups == null)
        {
          RF2DescriptionGroup[] temp = new RF2DescriptionGroup[this.descriptionGroupIds.Count];
          for (Int32 i = 0; i < this.descriptionGroupIds.Count; i++)
            temp[i] = this.Parser.GetDescriptionGroup(this.descriptionGroupIds[i]);
          this.descriptionGroups = temp;
        }
        return this.descriptionGroups;
      }
    }

    /// <summary>
    /// Description groups
    /// </summary>
    RF2DescriptionGroup[] descriptionGroups = null;

    /// <summary>
    /// Lazy load Source relationship groups.
    /// </summary>
    public RF2RelationshipGroup[] SourceRelationships
    {
      get
      {
        // # Not Tested.
        if (this.sourceRelationshipGroups == null)
        {
          RF2RelationshipGroup[] temp = new RF2RelationshipGroup[this.sourceRelationshipIds.Count];
          for (Int32 i = 0; i < this.sourceRelationshipIds.Count; i++)
            temp[i] = this.Parser.GetRelationshipGroup(this.sourceRelationshipIds[i]);
          this.sourceRelationshipGroups = temp;
        }
        return this.sourceRelationshipGroups;
      }
    }

    /// <summary>
    /// Relationship groups in which this concept is the source.
    /// </summary>
    RF2RelationshipGroup[] sourceRelationshipGroups = null;

    /// <summary>
    /// Lazy load Destination relationship groups.
    /// </summary>
    public RF2RelationshipGroup[] DestinationRelationships
    {
      get
      {
        // # Tested.
        if (this.destinationRelationshipGroups == null)
        {
          RF2RelationshipGroup[] temp = new RF2RelationshipGroup[this.destinationRelationshipIds.Count];
          for (Int32 i = 0; i < this.destinationRelationshipIds.Count; i++)
            temp[i] = this.Parser.GetRelationshipGroup(this.destinationRelationshipIds[i]);
          this.destinationRelationshipGroups = temp;
        }
        return this.destinationRelationshipGroups;
      }
    }

    /// <summary>
    /// Relationship groups in which this concept is the destination.
    /// </summary>
    RF2RelationshipGroup[] destinationRelationshipGroups = null;

    /// <summary>
    /// Constructor
    /// </summary>
    public RF2ConceptGroup()
    {
    }

    /// <summary>
    /// Add concept to concept group.
    /// </summary>
    /// <param name="concept"></param>
    public void AddConcept(RF2Concept concept)
    {
      this.items.Add(concept);
    }

    /// <summary>
    /// Add description group to concept group.
    /// </summary>
    /// <param name="descriptionGroup"></param>
    public void AddDescriptionGroup(RF2DescriptionGroup descriptionGroup)
    {
      if (this.descriptionGroupIds.Contains(descriptionGroup.Id) == true)
        return;
      this.descriptionGroupIds.Add(descriptionGroup.Id);
      // force lazy reloading of description groups id called.
      this.descriptionGroups = null;
    }

    /// <summary>
    /// Add a source relationship to concept group.
    /// </summary>
    /// <param name="relationshipGroup">Relationship in which this instance is the source</param>
    public void AddSourceRelationship(RF2RelationshipGroup relationshipGroup)
    {
      if (this.sourceRelationshipIds.Contains(relationshipGroup.Id) == true)
        return;
      this.sourceRelationshipIds.Add(relationshipGroup.Id);

      // force lazy reloading of source relationship groups id called.
      this.sourceRelationshipGroups = null;
    }

    /// <summary>
    /// Add a destination relationship to concept group.
    /// </summary>
    /// <param name="relationshipGroup">Relationship in which this instance is the destination</param>
    public void AddDestinationRelationship(RF2RelationshipGroup relationshipGroup)
    {
      if (this.destinationRelationshipIds.Contains(relationshipGroup.Id) == true)
        return;
      this.destinationRelationshipIds.Add(relationshipGroup.Id);

      // force lazy reloading of source relationship groups id called.
      this.destinationRelationshipGroups = null;
    }

    /// <summary>
    /// Return Snomed fully qualified name.
    /// </summary>
    /// <returns></returns>
    public String GetFullySpecifiedName()
    {
      String fsnAny = String.Empty;

      foreach (RF2DescriptionGroup dg in this.DescriptionGroups)
      {
        foreach (RF2Description description in dg.Items)
        {
          if (description.TypeId == RF2Parser.FullySpecifiedNameConceptId)
          {
            if (description.LanguageCode == "en")
              return description.Term;
            fsnAny = description.Term;
          }
        }
      }

      return fsnAny;
    }

    /// <summary>
    /// Return list of all SNOMED relationship groups that are IsA relationships, where this concept is 
    /// the source.
    /// </summary>
    /// <returns></returns>
    public List<RF2RelationshipGroup> IsAChildRelationships()
    {
      List<RF2RelationshipGroup> retVal = new List<RF2RelationshipGroup>();

      foreach (RF2RelationshipGroup g in this.DestinationRelationships)
      {
        RF2Relationship relationship = g.Active;
        if (
          (relationship.Active == true) &&
          (relationship.TypeId == RF2Parser.IsAConceptId)
          )
        {
          retVal.Add(g);
        }
      }
      return retVal;
    }


    /// <summary>
    /// Override ToString method.
    /// </summary>
    /// <returns></returns>
    public override string ToString()
    {
      return $"RF2ConceptGroup [{Id}] {GetFullySpecifiedName()}";
    }
  }
}

File: Java/SnomedQuery/Parser/RF2ConceptGroup.java
package Parser;

import java.util.ArrayList;

/**
 * <p>
 * Group of concepts, all with same SNOMED id.
 * </p>
 * 
 * @author Travis Lukach
 */
public class RF2ConceptGroup extends TRF2ItemGroup<RF2Concept> {
	/**
	 * <p>
	 * List of all snomed descriptions for this concept.
	 * </p>
	 */
	private ArrayList<Long> descriptionGroupIds = new ArrayList<Long>();

	/**
	 * <p>
	 * List of all snomed attribute concepts in which this concept is the source
	 * concept.
	 * </p>
	 */
	private ArrayList<Long> sourceRelationshipIds = new ArrayList<Long>();

	/**
	 * <p>
	 * List of all snomed attribute concepts in which this concept is the
	 * destination concept.
	 * </p>
	 */
	private ArrayList<Long> destinationRelationshipIds = new ArrayList<Long>();

	/**
	 * <p>
	 * Chain of is a parents.
	 * </p>
	 */
	private ArrayList<Long> isAParents = new ArrayList<Long>();

	/**
	 * <p>
	 * Concept id for the class that this snomed is an instance of.
	 * </p>
	 */
	private long classConceptId = -1;

	/**
	 * <p>
	 * Top level node that this concept group is descended from.
	 * </p>
	 */
	private long topLevelConceptId = -1;

	/**
	 * <p>
	 * Gets <i>topLevelConceptId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getTopLevelConceptId() {
		return topLevelConceptId;
	}

	/**
	 * <p>
	 * Sets <i>topLevelConceptId</i>.
	 * </p>
	 * 
	 * @param topLevelConceptId
	 */
	public void setTopLevelConceptId(long topLevelConceptId) {
		this.topLevelConceptId = topLevelConceptId;
	}

	/**
	 * <p>
	 * Gets <i>classConceptId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getClassConceptId() {
		return classConceptId;
	}
	/**
	 * <p>
	 * Sets <i>classConceptId</i>.
	 * </p>
	 * 
	 * @param classConceptId
	 */
	public void setClassConceptId(long classConceptId) {
		this.classConceptId = classConceptId;
	}
	/**
	 * Gets <i>isAParents</i>.
	 * 
	 * @return
	 */
	public ArrayList<Long> getIsAParents() {
		return isAParents;
	}
	/**
	 * <p>
	 * Sets <i>isAParents</i>.
	 * </p>
	 * 
	 * @param isAParents
	 */
	public void setIsAParents(ArrayList<Long> isAParents) {
		this.isAParents = isAParents;
	}

	/**
	 * <p>
	 * Lazy load Description groups.
	 * </p>
	 * 
	 * @return
	 */
	public final RF2DescriptionGroup[] getDescriptionGroups() {
		// # Tested.
		if (this.descriptionGroups == null) {
			RF2DescriptionGroup[] temp = new RF2DescriptionGroup[this.descriptionGroupIds
					.size()];
			for (int i = 0; i < this.descriptionGroupIds.size(); i++) {
				temp[i] = this.getParser()
						.getDescriptionGroup(this.descriptionGroupIds.get(i));
			}
			this.descriptionGroups = temp;
		}
		return this.descriptionGroups;
	}

	/**
	 * <p>
	 * Description groups
	 * </p>
	 */
	private RF2DescriptionGroup[] descriptionGroups = null;

	/**
	 * <p>
	 * Lazy load Source relationship groups.
	 * </p>
	 */
	public final RF2RelationshipGroup[] getSourceRelationships() {
		// # Not Tested.
		if (this.sourceRelationshipGroups == null) {
			RF2RelationshipGroup[] temp = new RF2RelationshipGroup[this.sourceRelationshipIds
					.size()];
			for (int i = 0; i < this.sourceRelationshipIds.size(); i++) {
				temp[i] = this.getParser().getRelationshipGroup(
						this.sourceRelationshipIds.get(i));
			}
			this.sourceRelationshipGroups = temp;
		}
		return this.sourceRelationshipGroups;
	}

	/**
	 * <p>
	 * Relationship groups in which this concept is the source.
	 * </p>
	 */
	private RF2RelationshipGroup[] sourceRelationshipGroups = null;

	/**
	 * <p>
	 * Lazy load Destination relationship groups.
	 * </p>
	 * 
	 * @return
	 */
	public final RF2RelationshipGroup[] getDestinationRelationships() {
		// # Tested.
		if (this.destinationRelationshipGroups == null) {
			RF2RelationshipGroup[] temp = new RF2RelationshipGroup[this.destinationRelationshipIds
					.size()];
			for (int i = 0; i < this.destinationRelationshipIds.size(); i++) {
				temp[i] = this.getParser().getRelationshipGroup(
						this.destinationRelationshipIds.get(i));
			}
			this.destinationRelationshipGroups = temp;
		}
		return this.destinationRelationshipGroups;
	}

	/**
	 * <p>
	 * Relationship groups in which this concept is the destination.
	 * </p>
	 */
	private RF2RelationshipGroup[] destinationRelationshipGroups = null;

	/**
	 * <p>
	 * Constructor.
	 * </p>
	 */
	public RF2ConceptGroup() {
	}

	/**
	 * <p>
	 * Add concept to concept group.
	 * </p>
	 * 
	 * @param concept
	 */
	public final void addConcept(final RF2Concept concept) {
		this.items.add(concept);
	}

	/**
	 * <p>
	 * Add description group to concept group.
	 * </p>
	 * 
	 * @param descriptionGroup
	 */
	public final void addDescriptionGroup(
			final RF2DescriptionGroup descriptionGroup) {
		if (this.descriptionGroupIds.contains(descriptionGroup.getId()) == true) {
			return;
		}
		this.descriptionGroupIds.add(descriptionGroup.getId());
		// force lazy reloading of description groups id called.
		this.descriptionGroups = null;
	}

	/**
	 * <p>
	 * Add a source relationship to concept group.
	 * </p>
	 * 
	 * @param relationshipGroup
	 *            Relationship in which this instance is the source
	 */
	public final void addSourceRelationship(
			final RF2RelationshipGroup relationshipGroup) {
		if (this.sourceRelationshipIds.contains(relationshipGroup.getId()) == true) {
			return;
		}
		this.sourceRelationshipIds.add(relationshipGroup.getId());

		// force lazy reloading of source relationship groups id called.
		this.sourceRelationshipGroups = null;
	}

	/**
	 * Add a destination relationship to concept group.
	 * 
	 * @param relationshipGroup
	 *            Relationship in which this instance is the destination
	 */
	public final void addDestinationRelationship(
			final RF2RelationshipGroup relationshipGroup) {
		if (this.destinationRelationshipIds
				.contains(relationshipGroup.getId()) == true) {
			return;
		}
		this.destinationRelationshipIds.add(relationshipGroup.getId());

		// force lazy reloading of source relationship groups id called.
		this.destinationRelationshipGroups = null;
	}

	/**
	 * Return Snomed fully qualified name.
	 * 
	 * @return
	 */
	public final String getFullySpecifiedName() {
		String fsnAny = "";

		for (RF2DescriptionGroup dg : this.getDescriptionGroups()) {
			for (RF2Description description : dg.getItems()) {
				if (description.getTypeId() == RF2Parser.fullySpecifiedNameConceptId) {
					if (description.getLanguageCode().equals("en")) {
						return description.getTerm();
					}
					fsnAny = description.getTerm();
				}
			}
		}

		return fsnAny;
	}

	/**
	 * <p>
	 * Return list of all SNOMED relationship groups that are IsA relationships,
	 * where this concept is the source.
	 * </p>
	 * 
	 * @return ArrayList of Relationship groups.
	 */
	public final ArrayList<RF2RelationshipGroup> isAChildRelationships() {
		ArrayList<RF2RelationshipGroup> retVal = new ArrayList<RF2RelationshipGroup>();

		for (RF2RelationshipGroup g : this.getDestinationRelationships()) {
			RF2Relationship relationship = g.getActive();
			if ((relationship.isActive() == true)
					&& (relationship.getTypeId() == RF2Parser.isAConceptId)) {
				retVal.add(g);
			}
		}
		return retVal;
	}

	/**
	 * <p>
	 * Override ToString method.
	 * </p>
	 * 
	 * @return
	 */
	@Override
	public final String toString() {
		return String.format("RF2ConceptGroup [%1$s] %2$s", getId(),
				getFullySpecifiedName());
	}

}